improver/lightning.py

```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+
+        Args:
+            cubes:
+                Cubes of CAPE, LIFTIDX, PWAT, CIN, APCP.
+            model_id_attr:
+                The name of the dataset attribute to be used to identify the source
+                model when blending data from different models.
+
+        Returns:
+            Cube of lightning data
+
+        Raises:
+            ValueError:
+                If one of the cubes is not found or doesn't match the other
+        """
+        cape, liftidx, pwat, cin, apcp = self._get_inputs(cubes)
+
+        # Regression equations require math on cubes with incompatible units, so strip data
+        templ = apcp.copy()
+        cape = cape.data
+        liftidx = liftidx.data
+        pwat = pwat.data
+        cin = cin.data
+        apcp = apcp.data / 1000 * 39.3701  # convert (kg m-3) to (inches)
+
+        # Regression equation when CAPE and APCP are greater than zero:
+        lprob = cape * apcp
+        lprob = 0.13 * np.log(lprob + 0.7) + 0.05
+
+        # If APCP is very low, a separate regression equation is used to predict lightning
+        # probability. The definition of “very low” is raised slightly when PWAT values are
+        # high. This is because the model often produces showery precipitation that doesn’t
+        # access the actual instability in very moist environments:
+        lprob_noprecip = cape / (cin + 100.0)
+        lprob_noprecip = 0.025 * np.log(lprob_noprecip + 0.31) + 0.03
+        apcp = apcp - (pwat / 1000)
+
+        lprob[np.where(apcp < 0.01)] = lprob_noprecip[np.where(apcp < 0.01)]
+
+        # If there is no CAPE but the atmosphere is “close” to unstable, lightning does sometimes
+        # occur, especially when heavy precipitation may have stabilized the atmosphere in the
+        # model. Unstable values of lifted index are positive here:
+        liftidx = liftidx + 4.0
+
+        liftidx[liftidx < 0] = 0
+
+        lprob[np.where(cape <= 0)] = (
+            0.2 * (liftidx[np.where(cape <= 0)] * apcp[np.where(cape <= 0)]) ** 0.5
+        )
+
+        # Finally, the probability of lightning is reduced when there is not much PWAT, since
+        # graupel cannot form required for the charging process. The probability is reduced as:
+        lprob[np.where(pwat < 20)] = lprob[np.where(pwat < 20)] * (
+            pwat[np.where(pwat < 20)] / 20.0
+        )
+
+        # Limit probabilities to 95% as that is as skillful as the regression equations could get:
+        lprob[lprob > 0.95] = 0.95
+
+        # Make percentage:
+        data = lprob * 100.0
+
+        cube = create_new_diagnostic_cube(
+            name="""20_km_lightning_probability_over_the_valid_time_of_the_accumulated_
+                precipitation""",
+            units="1",
+            template_cube=templ,
+            data=data.astype(FLOAT_DTYPE),
+            mandatory_attributes=generate_mandatory_attributes(
+                cubes, model_id_attr=model_id_attr
+            ),
+        )
```

Comment:
I think that I've got a few metadata queries.

1. I note that this output, following the precipitation accumulation input, has a time coordinate where the validity time is defined as a midpoint between the bounds. For period diagnostics, we always define the validity time to match the end bound, so I thought that I'd query this. Is the use of validity time as the midpoint for period diagnostics your preferred approach? I note that the forecast period also uses the midpoint of the bounds as the "point".
2. Currently this cube is a probability cube but doesn't have a threshold coordinate. Would it be possible to add a threshold coordinate, similar to the `number_of_lightning_flashes per_unit_area` coordinate in the `LightningFromCapePrecip` plugin? Perhaps something like `lightning_in_area` would work as a threshold coordinate name? Currently, this cube is a bit implicit about the fact that it is predicting the probability of _any_ lightning. 
3. The cube name is a bit overloaded at the moment. Our lightning probability diagnostic is named: "probability_of_number_of_lightning_flashes_per_unit_area_in_vicinity_above_threshold", so perhaps something like `probability_of_lightning_in_vicinity_above_threshold` would be better. As our lightning probability diagnostic is also within a particular area, we've added a `radius_of_vicinity` coordinate, which in your case would be `radius_of_vicinity: 20000 m`. Would it make sense to arrange the metadata like that from your perspective?

Reply:
With regard to the above comments:

1. I was unaware of the convention that the Met Office followed regarding validity time falling at the end of a period for diagnostics.  We don't have a preferred approach and are fine with either method.  This algorithm is written in a way so that the convention used for the accumulated precipitation input will be propagated through to the outputs, so it really should support either convention of the input data.  Regardless, I've updated the unit tests to follow the Met Office convention of having the validity time be at the end of the period. 
2. Yes, we can add a threshold coordinate. Given that the `LightningFromCapePrecip` plugin also is forecasting the probability of _any_ lightning, I'm going to use the exact same threshold coordinate name as what that plugin uses for consistency.
3. Got it.  I simplified the name as you suggested to `probability_of_lightning_in_vicinity_above_threshold` and added an auxiliary coordinate to note the vicinity over which this product is valid for 20 km.